File: tools-backend/schemas/arbitrage.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class ArbitrageCalculationRequest(BaseModel):
    """Request for arbitrage calculation"""

    comex_price_usd: float = Field(
        gt=0, description="COMEX price in USD per troy ounce"
    )
    mcx_price_inr: float = Field(gt=0, description="MCX price in INR per 10 grams")
    usdinr_rate: float = Field(gt=0, description="USD/INR exchange rate")
    import_duty_percent: float = Field(
        default=2.5, ge=0, le=100, description="Import duty percentage"
    )
    contract_size_grams: int = Field(
        default=10, gt=0, description="Quote unit in grams (10 for MCX Gold)"
    )


class FairValueResult(BaseModel):
    """Fair value calculation result"""

    comex_price_usd: float
    usdinr_rate: float
    price_per_gram_inr: float
    import_duty_percent: float
    fair_value_inr: float
    contract_size_grams: int


class ArbitrageMetrics(BaseModel):
    """Arbitrage opportunity metrics"""

    mcx_price: float
    fair_value: float
    premium: float
    premium_percent: float
    z_score: Optional[float] = None
    percentile: Optional[float] = None
    signal: str  # "strong_long", "long", "neutral", "short", "strong_short"


class ProfitAnalysis(BaseModel):
    """Profit potential analysis"""

    gross_profit: float
    brokerage: float
    exchange_fees: float
    tax: float
    total_costs: float
    net_profit: float
    net_profit_percent: float


class USDINRSensitivity(BaseModel):
    """USDINR sensitivity analysis"""

    current_usdinr: float
    new_usdinr: float
    usdinr_change: float
    current_fair_value: float
    new_fair_value: float
    fair_value_change: float
    fair_value_change_percent: float


class ArbitrageCalculationResponse(BaseModel):
    """Complete arbitrage analysis response"""

    timestamp: datetime
    symbol: str
    fair_value: FairValueResult
    arbitrage: ArbitrageMetrics
    profit_analysis: ProfitAnalysis
    recommendation: str
    risk_level: str  # "low", "medium", "high"


class ArbitrageHistoryRequest(BaseModel):
    """Request for historical arbitrage data"""

    symbol: str
    days: int = Field(default=30, ge=1, le=365)


class ArbitrageAlertRequest(BaseModel):
    """Request to set up arbitrage alert"""

    symbol: str
    alert_type: str = Field(description="premium_above, premium_below, signal_change")
    threshold: float = Field(description="Threshold value (e.g., premium % or z-score)")
    notification_channel: str = Field(
        default="telegram", description="telegram, email, or both"
    )

File: tools-frontend/lib/api/arbitrage.ts
import { apiClient } from "../api-client"
import type { ArbitrageRequest, ArbitrageResponse } from "@/types"

export const arbitrageApi = {
  calculate: async (data: ArbitrageRequest): Promise<ArbitrageResponse> => {
    return apiClient.post("/arbitrage/calculate", data)
  },

  getRealtime: async (commodity: string): Promise<{
    timestamp: string
    symbol: string
    fair_value: {
      comex_price_usd: number
      usdinr_rate: number
      price_per_gram_inr: number
      import_duty_percent: number
      fair_value_inr: number
      contract_size_grams: number
    }
    arbitrage: {
      mcx_price: number
      fair_value: number
      premium: number
      premium_percent: number
      signal: string
    }
    profit_analysis: {
      gross_profit: number
      net_profit: number
      net_profit_percent: number
    }
    recommendation: string
    risk_level: string
    data_sources: {
      comex: string
      mcx: string
      usdinr: string
    }
  }> => {
    return apiClient.get(`/arbitrage/realtime?symbol=${commodity}`)
  },

  getSensitivity: async (
    commodity: string,
    comexPriceUsd: number,
    currentUsdinr: number,
    usdinrChange: number,
    contractSize: number = 10
  ): Promise<{
    analysis: {
      current_usdinr: number
      new_usdinr: number
      usdinr_change: number
      current_fair_value: number
      new_fair_value: number
      fair_value_change: number
      fair_value_change_percent: number
    }
    interpretation: {
      impact_per_rupee: number
      direction: string
      magnitude: string
    }
  }> => {
    return apiClient.post(
      `/arbitrage/sensitivity?comex_price_usd=${comexPriceUsd}&current_usdinr=${currentUsdinr}&usdinr_change=${usdinrChange}&contract_size=${contractSize}`
    )
  },

  getHistory: async (
    symbol: string = "GOLD",
    days: number = 30,
    maxPoints: number = 500
  ): Promise<{
    symbol: string
    days: number
    data_count: number
    total_points?: number
    downsampled?: boolean
    data: Array<{
      recorded_at: string
      comex_price_usd: number
      mcx_price_inr: number
      usdinr_rate: number
      fair_value_inr: number
      premium: number
      premium_percent: number
      signal: string
      z_score: number | null
      percentile: number | null
    }>
    statistics: {
      average_premium_percent: number
      std_deviation: number
      min_premium_percent: number
      max_premium_percent: number
      signal_distribution: Record<string, number>
    } | null
    message?: string
  }> => {
    return apiClient.get(`/arbitrage/history?symbol=${symbol}&days=${days}&max_points=${maxPoints}`)
  },

  recordHistory: async (symbol: string = "GOLD"): Promise<{
    message: string
    id: number
    symbol: string
    recorded_at: string
    premium_percent: number
  }> => {
    return apiClient.post(`/arbitrage/history/record?symbol=${symbol}`)
  },
}
